sov-modules/sov-modules-impl/prover-incentives/src/call.rs

```diff
@@ -0,0 +1,170 @@
+            .transfer_from(prover, &self.address, coins, working_set)?;
+
+        // Update our record of the total bonded amount for the sender.
+        // This update is infallible, so no value can be destroyed.
+        let old_balance = self
+            .bonded_provers
+            .get(prover, working_set)
+            .unwrap_or_default();
+        let total_balance = old_balance + bond_amount;
+        self.bonded_provers.set(prover, total_balance, working_set);
+
+        // Emit the bonding event
+        response.add_event(
+            "bonded_prover",
+            &format!("new_deposit: {bond_amount:?}. total_bond: {total_balance:?}"),
+        );
+
+        Ok(response)
+    }
+
+    /// Try to bond the requested amount of coins from context.sender()
+    pub(crate) fn bond_prover(
+        &self,
+        bond_amount: u64,
+        context: &C,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<sov_modules_api::CallResponse> {
+        self.bond_prover_helper(bond_amount, context.sender(), working_set)
+    }
+
+    /// Try to unbond the requested amount of coins with context.sender() as the beneficiary.
+    pub(crate) fn unbond_prover(
+        &self,
+        context: &C,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<sov_modules_api::CallResponse> {
+        let mut response = CallResponse::default();
+
+        // Get the prover's old balance.
+        if let Some(old_balance) = self.bonded_provers.get(context.sender(), working_set) {
+            // Transfer the bond amount from the sender to the module's address.
+            // On failure, no state is changed
+            let coins = Coins {
+                token_address: self
+                    .bonding_token_address
+                    .get(working_set)
+                    .expect("Bonding token address must be set"),
+                amount: old_balance,
+            };
+            // Try to unbond the entire balance
+            // If the unbonding fails, no state is changed
+            self.bank
+                .transfer_from(&self.address, context.sender(), coins, working_set)?;
+
+            // Update our internal tracking of the total bonded amount for the sender.
+            self.bonded_provers.set(context.sender(), 0, working_set);
+
+            // Emit the unbonding event
+            response.add_event(
+                "unbonded_prover",
+                &format!("amount_withdrawn: {old_balance:?}"),
+            );
+        }
+
+        Ok(response)
+    }
+
+    /// Try to process a zk proof, if the prover is bonded.
+    pub(crate) fn process_proof(
+        &self,
+        proof: &[u8],
+        context: &C,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<sov_modules_api::CallResponse> {
+        let mut response = CallResponse::default();
+
+        // Get the prover's old balance.
+        if let Some(old_balance) = self.bonded_provers.get(context.sender(), working_set) {
```

Comment:
Whole logic of this method is inside this if, maybe we invert it with early return, so it is easier to read?